File: LeetCode/017.letter-combinations-of-a-phone-number/letter-combinations-of-a-phone-number.java
class Solution {
    public List<String> letterCombinations(String digits) {
        List<String>res = new ArrayList<>();
        if (digits.contains("1") || digits.length() == 0)   return res;
        String[] str = {"abc","def","ghi","jkl","mno","pqrs","tuv","wxyz"};
        func(str,digits,"",0,res);
        return res;
    }

    public void func(String[] str ,String digits,String string,int index,List<String>res){
        if (index >= digits.length()){
            res.add(string);
            return;
        }
        for (int i = 0;i < str[digits.charAt(index) - '2'].length();i++){
            func(str,digits,string + str[digits.charAt(index) - '2'].charAt(i),index + 1,res);
        }
    }
}

File: LeetCode/0017.letter-combinations-of-a-phone-number/letter-combinations-of-a-phone-number.cpp
class Solution {
public:
    vector<string> letterCombinations(string digits) {
        string str[] = {"abc", "def", "ghi", "jkl", "mno", "pqrs", "tuv", "wxyz"};
        vector<string>res;
        string value;
        if (digits.size() == 0) return res;
        helper(res, str, digits, 0, value);
        return res;
    }

    void helper(vector<string>&res, string str[], string digits, int index, string value){
        if (index >= digits.size()){
            res.push_back(value);
            return;
        }
        for (int i = 0; i < str[digits[index] - '2'].size(); i++){
            helper(res, str, digits, index + 1, value + str[digits[index] - '2'][i]);
        }
    }
};